### ISSUE TITLE
Import-related mdtests aren't isolated

### ISSUE THREAD

**Issue by dcreager** [astral-sh/ruff#15027]
When testing #15026 I got spurious test failures that look like the different test cases within an mdtest file aren't as isolated as they should be:

At commit 65da13900cd07c365159a67a6e1cda6442030e2f, the `Relative - Dotted` test fails when you test the entire file:

```
$ cargo test -p red_knot_python_semantic -- mdtest__import_relative
[snip]
relative.md - Relative - Dotted
  crates/red_knot_python_semantic/resources/mdtest/import/relative.md:39 unexpected error: [unresolved-import] "Cannot resolve import `.foo.bar.baz`"
  crates/red_knot_python_semantic/resources/mdtest/import/relative.md:41 unmatched assertion: revealed: Literal[42]
  crates/red_knot_python_semantic/resources/mdtest/import/relative.md:41 unexpected error: 1 [revealed-type] "Revealed type is `Unknown`"
```

but pass when you run that test by itself:

```
$ env MDTEST_TEST_FILTER="relative.md - Relative - Dotted" cargo test -p red_knot_python_semantic -- mdtest__import_relative
```

Renaming the files in that one test case to be unique within the mdtest file (c5fd3c64ae94e937fb3bd2c913830d8cb246df74) also causes test to pass for both invocations.

**Comment by MichaReiser** [comment]
I haven't figured out the failure yet but I think the test failing is the *expected* output because the `__init__.py` files are missing for the `bar` and `foo` package:

````
```toml
log = true
```

```py path=package/__init__.py
```

```py path=package/foo/__init__.py
```

```py path=package/foo/bar/__init__.py
```

```py path=package/foo/bar/baz.py
X = 42
```

```py path=package/bar.py
from .foo.bar.baz import X

reveal_type(X)  # revealed: Literal[42]
```
````
